drake/systems/plants/@Manipulator/parameterEstimation.m

```diff
@@ -90,103 +133,124 @@
-  qdd_data(:,i) = (H\(B*u_data(:,i) - C)) + .01*randn(nq,1);
 end
 
 s_data = sin(q_data);
 c_data = cos(q_data);
+dt_data = diff(t_data);
 
-ndata = length(t_data);
-M_data = reshape(msubs(M(:),[q;s;c;qd;qdd;u],[q_data;s_data;c_data;qd_data;qdd_data;u_data])',nq*ndata,nlp);
-Mb_data = reshape(msubs(Mb,[q;s;c;qd;qdd;u],[q_data;s_data;c_data;qd_data;qdd_data;u_data])',nq*ndata,1);
+% % Debugging the energy model
+% E1 = getmsspoly(T1+U1);
+% E = subs(E1,p,p_orig);
+% E_data = msubs(E,[q1;s1;c1;qd1;u;dt],[q_data;s_data;c_data;qd_data;u_data;t_data]);
+% diffE_data = diff(E_data);
+% dE1 = subs(dE,p,p_orig);
+% dE_data = msubs(dE1,[q1;qd1;q2;qd2;u;dt],[q_data(:,1:end-1);qd_data(:,1:end-1);...
+%     q_data(:,2:end);qd_data(:,2:end);u_data(:,1:end-1);dt_data]);
+% figure;plot(diffE_data,'-g'); hold on; plot(dE_data,'-r');
+% title(['True \DeltaE vs. Estimated \DeltaE']);
+% xlabel('Sample Time')
+% ylabel({'\DeltaE Magnitude' '(Joules)/sample'})
+% legend('True \DeltaE','Estimated \DeltaE');
 
-lp_est = -M_data\Mb_data;
-err_orig = M_data*lp_orig + Mb_data;
 
-err_est = M_data*lp_est + Mb_data;
-sqerr_orig = sum(err_orig'*err_orig);
-sqerr_est = sum(err_est'*err_est);
+if strcmp(options.model,'dynamic')
+    ndata = length(t_data);
+    variables = [q;s;c;qd;qdd;u];
+    data_variables = [q_data;s_data;c_data;qd_data;qdd_data;u_data];
+    M_data = reshape(msubs(M(:),variables,data_variables)',nq*ndata,nlp);
+    Mb_data = reshape(msubs(Mb,variables,data_variables)',nq*ndata,1);
+elseif strcmp(options.model,'energetic')
+    variables = [q1;s1;c1;qd1;q2;s2;c2;qd2;u;dt];
+    data_variables = [q_data(:,1:end-1);s_data(:,1:end-1);c_data(:,1:end-1);qd_data(:,1:end-1);...
+        q_data(:,2:end);s_data(:,2:end);c_data(:,2:end);qd_data(:,2:end);u_data(:,1:end-1);dt_data];
+    M_data = msubs(M(:),variables,data_variables)';
+    Mb_data = msubs(Mb,variables,data_variables)';
+end
 
-%keyboard;
+if strcmp(options.method,'nonlinprog') || strcmp(options.method,'linprog')
+    % Nonlinear least-squares solver
+    prog = NonlinearProgram(np);
+    prog=prog.addConstraint(BoundingBoxConstraint(pmin,pmax),1:np);
+    
+    if strcmp(options.method,'nonlinprog')
+        % % Only nonlinear least squares
+        % nonlinfun = @(x) nonlinerr(x,lp,p,M_data,Mb_data);
+        % prog=prog.addCost(FunctionHandleObjective(np,nonlinfun),1:np);
+    end
+    if strcmp(options.method,'linprog')
+        % Least squares -> Nonlinear least squares on lumped parameter solution
+        lp_est = -pinv(full(M_data))*Mb_data;
+        prog=prog.addQuadraticCost(eye(np),p_orig,1:np);
+        lpconstraint_handle = @(x) lpconstraint_fun(x,lp,p);
+        lpconstraint = FunctionHandleConstraint(lp_est,lp_est,nlp,lpconstraint_handle);
+        prog=prog.addConstraint(lpconstraint,1:np);
+    end
+    [x,F,info] = prog.solve(p_orig);
+    if(info ~= 1)
+    	error('failed to solve the problem');
+    end
+elseif strcmp(options.method,'lsqnonlin')
```

Comment:
we should support this as a fallback/option in the NonlinearProgram (perhaps we already do).  you shouldn't have to special case it here.